Implement Caballero et al. spectral factor model

Jacc0027 opened the pull request:
- [x] Closes #1278 
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

<!-- Brief description of the problem and proposed solution (if not already fully described in the issue linked to above): --> https://github.com/pvlib/pvlib-python/issues/1278

[AM_AOD_PW_spectral_correction.zip](https://github.com/pvlib/pvlib-python/files/7061393/AM_AOD_PW_spectral_correction.zip)

kanderso-nrel commented:
Hi @Jacc0027 -- sorry for the delay if you were waiting for feedback on this PR!  It looks great so far.  To make a more detailed review easier, could you address the code style issues noted by "Stickler CI" on this page?  https://github.com/pvlib/pvlib-python/pull/1296/files

This will also need some tests (see the [contributing docs](https://pvlib-python.readthedocs.io/en/stable/contributing.html#testing)) to check that the code works on all the various python versions and package versions we support.  Please reach out if you need input from us to move this PR forward :) And thanks again for the contribution!

Jacc0027 commented:
> Hi @Jacc0027 -- sorry for the delay if you were waiting for feedback on this PR! It looks great so far. To make a more detailed review easier, could you address the code style issues noted by "Stickler CI" on this page? https://github.com/pvlib/pvlib-python/pull/1296/files
> 
> This will also need some tests (see the [contributing docs](https://pvlib-python.readthedocs.io/en/stable/contributing.html#testing)) to check that the code works on all the various python versions and package versions we support. Please reach out if you need input from us to move this PR forward :) And thanks again for the contribution!

Hi @kanderso-nrel , thank you for your help.

After a bit of time and testing, I finally managed to adapt the code and pass the Stickler CI test.

It is a pleasure for me to contribute to this wonderful project. 

I remain at your disposal for any further issues.

kanderso-nrel commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,219 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+
+    if np.min(airmass_absolute) < 0.58:
+        warn('Exceptionally low air mass: ' +
+             'model not intended for extra-terrestrial use')
```
Question for the pvlib maintainers: `first_solar_spectral_correction` copied these warnings from the MATLAB implementation, but it seems a little out of place in pvlib-python in 2021.  Worth keeping?

cwhanse replied:
I don't think so.

kanderso-nrel reviewed (commented):
Thanks for the perseverance dealing with Stickler :) I think the next step is to add unit tests for this new function.  The tests for `first_solar_spectral_correction` are a good template to start from: https://github.com/pvlib/pvlib-python/blob/master/pvlib/tests/test_atmosphere.py#L88-L149

Together, the tests should run every line of code in the new function.  Any lines of code not covered by the tests will be marked by "Codecov" in the same place that Stickler was commenting. 

Also, can you add an entry in `/docs/sphinx/source/api.rst` for your new function?  Next to `first_solar_spectral_correction` probably makes sense.

I put some minor comments below and will do a more thorough review later.  Thanks again for this PR!

A note for the pvlib maintainers: it might be time to move the existing spectral modifier functions into `pvlib.spectrum` as mentioned in https://github.com/pvlib/pvlib-python/pull/1062#issuecomment-696220404.  But to keep this PR simple I propose we leave the new function in `atmosphere.py` for now and move it and the existing functions in a follow-on PR.

Jacc0027 commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,219 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+        maximum atmospheric precipitable water. Any pw value higher than max_pw
+        is set to NaN to avoid model divergence. [cm]
+
+    module_type : None or string, default None
```
Done, thanks for the recommendation

Jacc0027 commented:
Hi @kanderso-nrel, I reply your messages below.

"Thanks for the perseverance dealing with Stickler :) I think the next step is to add unit tests for this new function. The tests for `first_solar_spectral_correction` are a good template to start from: https://github.com/pvlib/pvlib-python/blob/master/pvlib/tests/test_atmosphere.py#L88-L149"

Attached is a text document in which I have tried to adapt the code of the new function to the pre-existing first solar tests. I have tried to verify the tests by running them locally according to "https://pvlib-python.readthedocs.io/en/latest/contributing.html" but I have not succeeded. I would appreciate help to manage the execution of these tests.
[Test spectral corrections_AM_AOD_PW.txt](https://github.com/pvlib/pvlib-python/files/7273190/Test.spectral.corrections_AM_AOD_PW.txt)

"Together, the tests should run every line of code in the new function. Any lines of code not covered by the tests will be marked by "Codecov" in the same place that Stickler was commenting."

I have doubts as to whether this will be solved once the above tests have been solved or whether a separate action should be taken. Tthe problem is that codecov is not running on all code and I don't know how to change it.
 
"Also, can you add an entry in `/docs/sphinx/source/api.rst` for your new function? Next to `first_solar_spectral_correction` probably makes sense."
It has been already requested.

Sorry for the delay in replying, unfortunately I can't devote myself to research and my work has me quite absorbed. With respect to the doubts, the truth is that I had always programmed in Matlab and this is posing a challenge for me.

thank you again for your time and help.

Jacc0027 commented:
![image](https://user-images.githubusercontent.com/88579674/139541319-daceb1b4-15ff-4309-9b63-4ba636989cef.png)

Please, can you help me solve the import of atmosphere.py?
With this I will be able to advance in the development of the tests. 

Thank you very much in advance.

kanderso-nrel commented:
Hi @Jacc0027 -- I apologize for not replying earlier; somehow I missed your previous message.


For the test file you linked in this comment (https://github.com/pvlib/pvlib-python/pull/1296#issuecomment-932930650), I get an error `TypeError: 'tuple' object is not callable`.  This is because the `asi` tuple defined on lines 29 to 32 in that file does not have a comma after it.  So it just needs a comma to fix it:

```
    ('asi', np.array(
        [[ 1.0555275 ,  0.87707583,  0.72243772],
         [ 1.11225204,  0.93665901,  0.78487953],
         [ 1.14555295,  0.97084011,  0.81994083]]))    # <-- put a comma here
```

However I'm not sure what to suggest about that screenshot.  That code seems ok to me.  What error message is being reported?  

> the truth is that I had always programmed in Matlab and this is posing a challenge for me.

I think many of us have experienced the same :)  For the tests, if pytest is not making any sense, you can just assemble a set of inputs and expected outputs, and I am willing to get it working with pytest.  But I don't know what the "right answer" should be so I would need you to provide them.

Jacc0027 commented:
Hi @kanderso-nrel, I'm glad to read you again.

> Hi @Jacc0027 -- I apologize for not replying earlier; somehow I missed your previous message.
> 
Don't worry, as the saying goes, "It's better late than never".

> For the test file you linked in this comment ([#1296 (comment)](https://github.com/pvlib/pvlib-python/pull/1296#issuecomment-932930650)), I get an error `TypeError: 'tuple' object is not callable`. This is because the `asi` tuple defined on lines 29 to 32 in that file does not have a comma after it. So it just needs a comma to fix it:
> 
> ```
>     ('asi', np.array(
>         [[ 1.0555275 ,  0.87707583,  0.72243772],
>          [ 1.11225204,  0.93665901,  0.78487953],
>          [ 1.14555295,  0.97084011,  0.81994083]]))    # <-- put a comma here
> ```
> Thanks for the review.

> However I'm not sure what to suggest about that screenshot. That code seems ok to me. What error message is being reported?
> 
What happens is that it doesn't let me import code from other files. Maybe I should change IDE to program in python. I am using Visual Studio Code. Do you have any suggestions based on your experience?

> > the truth is that I had always programmed in Matlab and this is posing a challenge for me.
> 
> I think many of us have experienced the same :) For the tests, if pytest is not making any sense, you can just assemble a set of inputs and expected outputs, and I am willing to get it working with pytest. But I don't know what the "right answer" should be so I would need you to provide them.

Let's see if I can help you with this. I attach a compressed file (ZIP) containing two files. One programmed in python that works independently of PVlib and with which the expected results can be obtained. So, that results can be put in the tests file for each type of module based on AM, AOD and PW values.
The second file is in csv format and in it we must write in order, and separated by commas, the values ​​of AM, AOD and PW on which we want to calculate the value of MM through the previous file. I have already included a few examples.

In the meantime, I will research other IDEs that allow me to program in python and let me import data from different scripts. Thank you very much for everything, we keep in touch.

[AM-AOD-PW_spectral_correction_data_test.zip](https://github.com/pvlib/pvlib-python/files/7449332/AM-AOD-PW_spectral_correction_data_test.zip)

kanderso-nrel commented:
> What happens is that it doesn't let me import code from other files. Maybe I should change IDE to program in python. I am using Visual Studio Code. Do you have any suggestions based on your experience?

The IDE should not really matter, unless perhaps it is misconfigured somehow.  For code like `from pvlib import atmosphere` to succeed, `pvlib` needs to be accessible from [sys.path](https://docs.python.org/3/library/sys.html#sys.path).  You can achieve that in a few ways:
1) Install your modified version of `pvlib` using `pip install /path/to/your/local/pvlib`.  This will install pvlib into your `site-packages` folder, which should be one of the entries in `sys.path`. 
2) Or, you can set the [current working directory](https://docs.python.org/3/library/os.html#os.getcwd) to be the same as where your local copy of pvlib is.

I hope one of those fixes the import issue, but if it still isn't working, it's okay to use the automatic tests on this page (e.g. "Test_bare_Linux Python38") to see if the tests are working.  

Anyway I have some edits in https://github.com/Jacc0027/pvlib-python/pull/2 with the tests based on your file, and a few other edits.

Jacc0027 commented:
> > What happens is that it doesn't let me import code from other files. Maybe I should change IDE to program in python. I am using Visual Studio Code. Do you have any suggestions based on your experience?
> 
> The IDE should not really matter, unless perhaps it is misconfigured somehow. For code like `from pvlib import atmosphere` to succeed, `pvlib` needs to be accessible from [sys.path](https://docs.python.org/3/library/sys.html#sys.path). You can achieve that in a few ways:
> 
> 1. Install your modified version of `pvlib` using `pip install /path/to/your/local/pvlib`.  This will install pvlib into your `site-packages` folder, which should be one of the entries in `sys.path`.
> 2. Or, you can set the [current working directory](https://docs.python.org/3/library/os.html#os.getcwd) to be the same as where your local copy of pvlib is.
> 
Thanks, the truth is that I had tried to configure the path from sys.path to pvlib folder but it still did not work.

> I hope one of those fixes the import issue, but if it still isn't working, it's okay to use the automatic tests on this page (e.g. "Test_bare_Linux Python38") to see if the tests are working.
> 
> Anyway I have some edits in [Jacc0027#2](https://github.com/Jacc0027/pvlib-python/pull/2) with the tests based on your file, and a few other edits.

Thanks, I have reviewed and merged the changes into this PR, it has been much more simplified and clear. 
Then, I see that some tests are not executed. What can we do in order to solve it?

mikofski commented:
There are some docs you can find on setting up your python environment using either conda or the building python venv library, then the trick is to activate that environment and install pvlib in develop mode. Follow the [installation instructions in the pvlib docs](https://pvlib-python.readthedocs.io/en/stable/installation.html#install-as-an-editable-library). Then from any ide you can select that virtual end as your interpreter

kanderso-nrel commented:
> Then, I see that some tests are not executed. What can we do in order to solve it?

This was caused by an unrelated problem and is solved in #1335.  If you merge pvlib/master into your branch it will fix it here too.  You can use git commands to do that merge, or you can use a GitHub pull request to do the merge with this link: https://github.com/Jacc0027/pvlib-python/compare/Spectral-corrections...pvlib:master

Jacc0027 commented:
> There are some docs you can find on setting up your python environment using either conda or the built in python `venv` library, then the trick is to activate that environment and install pvlib in develop mode. Follow the [installation instructions in the pvlib docs](https://pvlib-python.readthedocs.io/en/stable/installation.html#install-as-an-editable-library). Then from any ide you can select that virtual end as your interpreter

Thank you so much @mikofski ! I will inquire about it as soon as possible.

Jacc0027 commented:
I definitely give up for today. There came a time when I only got one line of code not covered by codecov. I think the reason is that it never occurs the case that "module type" and "Coefficients" are different from "None" in the test file. 
I have updated the values ​​of the tests and in atmosphere.py so that the values ​​of aod_ref and pw_ref are 0.084 and 1.42 respectively. Without knowing how, I get an error in all the tests and I don't know why.

Can anyone help me to solve it? Thank you very much in advance.

kanderso-nrel commented:
> I think the reason is that it never occurs the case that "module type" and "Coefficients" are different from "None" in the test file.

For this you can define another test function and put it in `test_atmosphere.py` with the others, something like this:

```python
def test_AM_AOD_PW_spectral_correction_supplied_ambiguous():
    dummy_coeffs = (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1)
    with pytest.raises(TypeError):
        atmosphere.AM_AOD_PW_spectral_correction(1, 1, 1, module_type='cdte', coefficients=dummy_coeffs)
```

For the failing tests, the immediate problem is that the inputs (AM, PW, AOD) are of length 6, and therefore so is the output MM array.  But that length 6 array is then compared with the "expected" array, which has length 5.  So either the "expected" arrays are missing an element, or the inputs have an extra element.  But it looks like there is still a difference in values in addition to the length difference.

Jacc0027 commented:
> > I think the reason is that it never occurs the case that "module type" and "Coefficients" are different from "None" in the test file.
> 
> For this you can define another test function and put it in `test_atmosphere.py` with the others, something like this:
> 
> ```python
> def test_AM_AOD_PW_spectral_correction_redundant():
>     dummy_coeffs = (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1)
>     with pytest.raises(TypeError):
>         atmosphere.AM_AOD_PW_spectral_correction(1, 1, 1, module_type='cdte', coefficients=dummy_coeffs)
> 
Magnificent @kanderso-nrel, you are amazing!! It worked the first time!
> 
> For the failing tests, the immediate problem is that the inputs (AM, PW, AOD) are of length 6, and therefore so is the output MM array. But that length 6 array is then compared with the "expected" array, which has length 5. So either the "expected" arrays are missing an element, or the inputs have an extra element. But it looks like there is still a difference in values in addition to the length difference.

You were right. I was missing one of the expected output data. I have modified it and now everything is correct.

I am finally understanding how Github and tests work. This is mainly due to the great help received. By the way, if anyone has contacts in the Guinness book of records, tell them that I have achieved the highest ratio of Commits per hour in Github! :)

Now we only have 1 error. "codecov / project / library". After reviewing it, I can see errors in the next file: "test.forecast.py".  What can we do in this case?

cwhanse commented:
> I can see errors in the next file: "test.forecast.py". What can we do in this case?

You can ignore that.  Please assign me to review when this PR is ready.

Jacc0027 commented:
> > I can see errors in the next file: "test.forecast.py". What can we do in this case?
> 
> You can ignore that. Please assign me to review when this PR is ready.

Thanks @cwhanse . I believe that all the steps have already been completed before ending this PR.
Also, note that I already requested your review a few days ago when I was stuck and it does not allow me to do it again.

I remain at your disposal in case I had to do something additional.

adriesse commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,153 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
     return - np.log(aod1 / aod2) / np.log(lambda1 / lambda2)
+
+
+def AM_AOD_PW_spectral_correction(airmass_absolute, aod500, pw,
```
Would it be preferable to us the main author's name in the function name?

Jacc0027 replied:
Ok, taking into account the authors' initials and in order to make the pronunciation of the name easy (as far as possible), the name I propose is: jaemfag_spectral_corrections

Jacc0027 commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,153 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+    the effect on module short circuit current of variation in the
+    spectral irradiance, :math:`MM` is estimated from absolute
+    (pressure-adjusted) AM, :math:`ama`, AOD at 500 nm, :math:`aod500`
+    and PW, :math:`pw`.
```
Ok, we can implement it.

Jacc0027 commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,153 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+    elif module_type is None and coefficients is not None:
+        pass
+    elif module_type is None and coefficients is None:
+        raise TypeError('No valid input provided, both module_type'
```
Accepted. I must check the rest of the code to see that there is no conflict.

Jacc0027 commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,153 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+        raise TypeError('No valid input provided, both module_type'
+                        + 'and coefficients are None')
+    else:
+        raise TypeError('Cannot resolve input, must supply only one'
```
Great @cwhanse!, if this is an advantage for the user, we will implement it. Thank you very much!

PS: I hope to implement the changes later this week, in the meantime, any suggestions for improvement are still welcome.

cwhanse commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,151 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+
+    Estimates a spectral mismatch modifier :math:`M` representing
+    the effect on module short circuit current of variation in the
+    spectral irradiance, :math:`MM` is estimated from absolute
```
```suggestion
    spectral irradiance. :math:`M` is estimated from absolute
```

cwhanse reviewed (commented):
Mostly fixes typos. Can you add descriptions to the other two arguments and fix the merge conflict in the whatsnew file?

Jacc0027 commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,151 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+        0.0583, -0.0159, 0.01251, 0.0109, 1, 0)
+
+    if module_type is None and coefficients is None:
+        raise TypeError('Invalid input provided, both module_type'
```
Updated

Jacc0027 commented:
> Mostly fixes typos. Can you add descriptions to the other two arguments and fix the merge conflict in the whatsnew file?

Hi Cliff, I guess I've done what you've requested, but am still receiving a conflict regarding to the following file: docs/sphinx/source/api.rst

and the next message:
Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.

can you help me to solve this issue?

Thanks in advance

cwhanse commented:
@Jacc0027 the merge conflict doesn't show up here, so I think you need to merge the current `main` from pvlib to see what's going on.

If you have `upstream` pointed to pvlib-python/main, then checkout this branch and

```
git pull upstream main
```

I hope you'll see a git message that you have a conflicted api.rst file and then can fix it.

Jacc0027 commented:
> @Jacc0027 the merge conflict doesn't show up here, so I think you need to merge the current `main` from pvlib to see what's going on.
> 
> If you have `upstream` pointed to pvlib-python/main, then checkout this branch and
> 
> ```
> git pull upstream main
> ```
> 
> I hope you'll see a git message that you have a conflicted api.rst file and then can fix it.

I noticed that the api.rst file did not exist in the main repository, so I have proposed to delete it and there are no more conflicts in this sense.

cwhanse commented:
@Jacc0027 that fixed the conflicts. Looks like some tests need corrections.

Jacc0027 commented:
> @Jacc0027 that fixed the conflicts. Looks like some tests need corrections.

Hi, I see..., they weren't appearing before. I'll take a look. Regards,

Jacc0027 commented:
> @Jacc0027 that fixed the conflicts. Looks like some tests need corrections.

I've proposed a couple of changes within the test file. Let's see if something has been solved.

kandersolar commented:
Thanks @Jacc0027 for the effort here!  I think I see how to get the tests working; would you mind if I pushed some commits to this PR for that?  Then after a bit of documentation I think this PR will be ready.

Jacc0027 commented:
> Thanks @Jacc0027 for the effort here! I think I see how to get the tests working; would you mind if I pushed some commits to this PR for that? Then after a bit of documentation I think this PR will be ready.

Hi @kandersolar , no problem. Go ahead with the solution. Thanks for the support!

kandersolar commented:
Looks like things are working!  I think these are the last main things, @Jacc0027 can you take a look:
- Fill in the two "TODO" parameter descriptions 
- Add an entry in `/docs/sphinx/source/reference/airmass_atmospheric.rst` for this new function
- Add a new item in `/docs/sphinx/source/whatsnew/v0.9.6.rst` for this enhancement, and yourself as a contributor at the bottom

Jacc0027 commented:
> Looks like things are working! I think these are the last main things, @Jacc0027 can you take a look:
> 
> * Fill in the two "TODO" parameter descriptions
> * Add an entry in `/docs/sphinx/source/reference/airmass_atmospheric.rst` for this new function
> * Add a new item in `/docs/sphinx/source/whatsnew/v0.9.6.rst` for this enhancement, and yourself as a contributor at the bottom

Thanks @kandersolar you did it!. I have done the 3 things you've requested me. Please, let me know if you need something else.

kandersolar commented:
Hmm, for some reason this PR is not showing any changes to the `reference` and `whatsnew` pages: https://github.com/pvlib/pvlib-python/pull/1296/files

Maybe you only changed them locally but did not include them in a commit?

Jacc0027 commented:
> Hmm, for some reason this PR is not showing any changes to the `reference` and `whatsnew` pages: https://github.com/pvlib/pvlib-python/pull/1296/files
> 
> Maybe you only changed them locally but did not include them in a commit?

I requested to update both files after modifying them via Github online. After requesting the Commit of the changes, I got an option to create a PR for them. Should I create these new PRs?

kandersolar commented:
It would be better to include them in this PR so that all the changes are in one place.  If you are editing them with Github online, then I think you can add the changes to this PR by editing the files on the specific git branch associated with this PR (the `Spectral-corrections` branch in your fork of pvlib).  For example this is the link for the "what's new" file on the PR branch: https://github.com/Jacc0027/pvlib-python/blob/Spectral-corrections/docs/sphinx/source/whatsnew/v0.9.6.rst

If you edit that file and go to commit the change, it should give you the option "Commit directly to the `Spectral-corrections` branch", and then I think the commit will show up in this PR.

adriesse reviewed (commented):
I have two very old comments that show up as pending.  Does that mean nobody else saw them?  I'm going to click on submit now but will take another look when I get back to the office next week.

Jacc0027 commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,153 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+    as a function of AM, AOD and PW was performed according to [2] and [3].
+    As such, the polynomial adjustment coefficients included in [3]
+    were obtained.
+
```
Hi @adriesse, the reason for using synthetic spectra, generated through SMARTS2, was to ascertain the ideal polynomial type of the experimental equations related to the AM, AOD and PW parameters.

Jacc0027 commented on pvlib/atmosphere.py:
```diff
@@ -683,3 +683,153 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+    Spain for one year synchronously with both, broadband and
+    spectroradiometric measurements of 30º tilted global irradiance
+    south-facing logged in 5-min intervals. AM was estimated through
+    elevation data.
```
Hi @adriesse, yes it does. The solar elevation was measured by means of a solar spectral irradiance meter (SolarSIM-D2) from Spectrafy Inc. Then the AM was computed by using the sun’s zenith angle (z). Reference: F. Kasten and A. T. Young, “Revised optical air mass tables and approxi-mation formula,” Appl. Opt., vol. 28, pp. 4735–4738, Nov. 15, 1989.

adriesse replied:
My thinking was that your responses might make it into the doc string at some point...

Jacc0027 commented:
> It would be better to include them in this PR so that all the changes are in one place. If you are editing them with Github online, then I think you can add the changes to this PR by editing the files on the specific git branch associated with this PR (the `Spectral-corrections` branch in your fork of pvlib). For example this is the link for the "what's new" file on the PR branch: https://github.com/Jacc0027/pvlib-python/blob/Spectral-corrections/docs/sphinx/source/whatsnew/v0.9.6.rst
> 
> If you edit that file and go to commit the change, it should give you the option "Commit directly to the `Spectral-corrections` branch", and then I think the commit will show up in this PR.

Hi @kandersolar, done. Thank you for the guidelines!

adriesse commented on pvlib/atmosphere.py:
```diff
@@ -698,3 +698,153 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+
+    # Evaluate Spectral Shift
+    ama = airmass_absolute
+    modifier = (
+        coeff[0] + (ama) * coeff[1] + (ama * ama) * coeff[2]
+        + (ama * ama * ama) * coeff[3] + (ama * ama * ama * ama) * coeff[4]
+        + (aod500 - aod500_ref) * coeff[5]
+        + ((aod500 - aod500_ref) * (ama) * coeff[6]) * coeff[10]
+        + ((aod500 - aod500_ref) * (np.log(ama)) * coeff[6]) * coeff[11]
+        + (aod500 - aod500_ref) * (ama * ama) * coeff[7]
+        + (pw - pw_ref) * coeff[8] + (pw - pw_ref) * (np.log(ama)) * coeff[9])
```
This calculation is neither easy to read nor efficient.  I suggest you rewrite in three (like in the paper) efficient subcalculations and return their sum.

Also add code comments to explain the extra coefficients.

Jacc0027 replied:
Hi @adriesse , this is the way I know to calculate the spectral modifier index. Please, feel free to propose another set of equations to do it and we can commit them. Thanks!

Jacc0027 replied:
Ok, we can add code comments to explain the extra coefficients. The reason to add it was to unify the number of coefficients to be applied for each photovoltaic technology. Then, the extra coefficient is set at 0 or 1 to take into account the coefficient when needed.

adriesse commented on pvlib/atmosphere.py:
```diff
@@ -698,3 +698,153 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+    else:
+        coeff = coefficients
+
+    # Evaluate Spectral Shift
```
```suggestion
    # Evaluate spectral correction factor
```

adriesse commented on pvlib/atmosphere.py:
```diff
@@ -698,3 +698,153 @@ def angstrom_alpha(aod1, lambda1, aod2, lambda2):
+
+def caballero_spectral_correction(airmass_absolute, aod500, pw,
+                                  module_type=None, coefficients=None,
+                                  aod500_ref=0.084, pw_ref=1.42):
```
I would personally take out the last two parameters.

Jacc0027 replied:
Hi @adriesse, I understand that the last 2 parameters are needed, they are constant values, so we must define them as inputs variables or inside the function.

Jacc0027 commented:
> Hi @Jacc0027, I just added a few changes. The main one is to move the new function to `pvlib.spectrum`, since we are consolidating the spectral modifier functions to that module (see #1628). I also made some changes to the function to better match typical pvlib style, and address @adriesse's comments above.
> 
> I think this PR is ready to merge, but I'll leave it open for a few days to allow a chance for anyone else to review again.

Thank you @kandersolar .

> Hi @Jacc0027, I just added a few changes. The main one is to move the new function to `pvlib.spectrum`, since we are consolidating the spectral modifier functions to that module (see #1628). I also made some changes to the function to better match typical pvlib style, and address @adriesse's comments above.
> 
> I think this PR is ready to merge, but I'll leave it open for a few days to allow a chance for anyone else to review again.

Hi @kandersolar, thank you so much for the effort! There have been two major solar energy events during the last week: Intersolar Europe 2023 and PVSC50. Perhaps during the week ahead there will be some additional comments. We'll see.

adriesse commented on pvlib/spectrum/mismatch.py:
```diff
@@ -446,3 +446,123 @@ def spectral_factor_sapm(airmass_absolute, module):
+    aod500_ref = 0.084
+    pw_ref = 1.4164
+
+    f_AM = (
+        coeff[0]
+        + coeff[1] * ama
+        + coeff[2] * ama**2
+        + coeff[3] * ama**3
+        + coeff[4] * ama**4
+    )
+    # Eq 6, with Table 1
+    f_AOD = (aod500 - aod500_ref) * (
+        coeff[5]
+        + coeff[10] * coeff[6] * ama
+        + coeff[11] * coeff[6] * np.log(ama)
+        + coeff[7] * ama**2
+    )
+    # Eq 7, with Table 1
+    f_PW = (precipitable_water - pw_ref) * (
+        coeff[8]
+        + coeff[9] * np.log(ama)
+    )
```
There are a several options for making this more efficient: Horner's method, np.polyval, pre-calculating ama powers, expanding **3 and **4, adding coefficients before multiplying where applicable. Just thought I'd mention that since it came up in other places in the past.  They would all reduce the much-appreciated code clarity to varying degrees, so I'll stop short of making a recommendation.

kandersolar replied:
I felt the same conflict between code clarity and efficiency.  I settled in favor of clarity after seeing that this inefficient implementation is still plenty fast by my standards (computes ~10 million values in one second).

Still, it does offend the sensibilities to calculate `x**3` on one line and `x**4` on the next :(

kandersolar commented:
Thanks again @Jacc0027 for this contribution, and my apologies that it took nearly two years to get it merged.  Better late than never, hopefully :)

Jacc0027 commented:
> 

Thank you so much @kandersolar, @adriesse, @cwhanse, @mikofski  for your support during this time! To a lesser or greater extent, the closing of this contribution has also been possible thanks to your assistance. I would really wish I could have made this contribution in other circumstances with more time available.
See you!
